pkg/controller/core/admissioncheck_controller.go

```diff
@@ -0,0 +1,230 @@
+	cqUpdateCh chan event.GenericEvent
+	watchers   []AdmissionCheckUpdateWatcher
+}
+
+func NewAdmissionCheckReconciler(
+	client client.Client,
+	qMgr *queue.Manager,
+	cache *cache.Cache,
+) *AdmissionCheckReconciler {
+	return &AdmissionCheckReconciler{
+		log:        ctrl.Log.WithName("admissioncheck-reconciler"),
+		qManager:   qMgr,
+		client:     client,
+		cache:      cache,
+		cqUpdateCh: make(chan event.GenericEvent, updateChBuffer),
+	}
+}
+
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=admissionchecks,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=admissionchecks/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=admissionchecks/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *AdmissionCheckReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ac := &kueue.AdmissionCheck{}
+
+	if err := r.client.Get(ctx, req.NamespacedName, ac); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log := log.FromContext(ctx).WithValues("admissionCheck", klog.KObj(ac))
+	if ac.DeletionTimestamp.IsZero() {
+		if controllerutil.AddFinalizer(ac, kueue.ResourceInUseFinalizerName) {
+			if err := r.client.Update(ctx, ac); err != nil {
+				return ctrl.Result{}, err
+			}
+			log.V(5).Info("Added finalizer")
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(ac, kueue.ResourceInUseFinalizerName) {
+			if cqs := r.cache.ClusterQueuesUsingAdmissionCheck(ac.Name); len(cqs) != 0 {
+				log.V(3).Info("admissionCheck is still in use", "ClusterQueues", cqs)
+				// We avoid to return error here to prevent backoff requeue, which is passive and wasteful.
+				// Instead, we drive the removal of finalizer by ClusterQueue Update/Delete events
+				// when the admissionCheck is no longer in use.
+				return ctrl.Result{}, nil
+			}
+			controllerutil.RemoveFinalizer(ac, kueue.ResourceInUseFinalizerName)
+			if err := r.client.Update(ctx, ac); err != nil {
+				return ctrl.Result{}, err
+			}
+			log.V(5).Info("Removed finalizer")
+		}
+	}
+	return ctrl.Result{}, nil
+}
+func (r *AdmissionCheckReconciler) notifyWatchers(oldAc, newAc *kueue.AdmissionCheck) {
+	for _, w := range r.watchers {
+		w.NotifyAdmissionCheckUpdate(oldAc, newAc)
+	}
+}
+
+func (r *AdmissionCheckReconciler) AddUpdateWatchers(watchers ...AdmissionCheckUpdateWatcher) {
+	r.watchers = append(r.watchers, watchers...)
+}
+
+func (r *AdmissionCheckReconciler) Create(e event.CreateEvent) bool {
+	ac, isAc := e.Object.(*kueue.AdmissionCheck)
+	if !isAc {
+		return false
+	}
+	defer r.notifyWatchers(nil, ac)
+	r.log.WithValues("admissionCheck", klog.KObj(ac)).V(5).Info("Create event")
+	if cqNames := r.cache.AddOrUpdateAdmissionCheck(ac); len(cqNames) > 0 {
+		r.qManager.QueueInadmissibleWorkloads(context.Background(), cqNames)
```

Comment:
Why do we need to move inadmissible workloads into the heap for every AdmissionCheck CRUD event?

Reply:
Ah, correct. We must move the inadmissible workloads into the heap here since the cq status changed from pending to active.